cpu/stm32f4/periph/adc.c

```diff
@@ -22,164 +22,63 @@
-
-typedef struct {
-    int max_value;
-} adc_config_t;
-
-adc_config_t adc_config[ADC_NUMOF];
-
-int adc_init(adc_t dev, adc_precision_t precision)
+static inline ADC_TypeDef *dev(adc_t line)
 {
-    ADC_TypeDef *adc = 0;
-
-    adc_poweron(dev);
-
-    switch (dev) {
-#if ADC_0_EN
-        case ADC_0:
-            adc = ADC_0_DEV;
-            ADC_0_PORT_CLKEN();
-            ADC_0_PORT->MODER |= (3 << (ADC_0_CH0_PIN * 2) | 3 << (ADC_0_CH1_PIN * 2));
-            break;
-#endif
-#if ADC_1_EN
-        case ADC_1:
-            adc = ADC_1_DEV;
-            ADC_1_PORT_CLKEN();
-            ADC_1_PORT->MODER |= (3 << (ADC_1_CH0_PIN * 2) | 3 << (ADC_1_CH1_PIN * 2));
-            break;
-#endif
-        default:
-            return -1;
-    }
-
-    /* reset control registers */
-    adc->CR1 = 0;
-    adc->CR2 = 0;
-    adc->SQR1 = 0;
-
-    /* set precision */
-
-    switch (precision) {
-        case ADC_RES_6BIT:
-            adc->CR1 |= ADC_CR1_RES_0 | ADC_CR1_RES_1;
-            adc_config[dev].max_value = 0x3f;
-            break;
-        case ADC_RES_8BIT:
-            adc->CR1 |= ADC_CR1_RES_1;
-            adc_config[dev].max_value = 0xff;
-            break;
-        case ADC_RES_10BIT:
-            adc->CR1 |= ADC_CR1_RES_0;
-            adc_config[dev].max_value = 0x3ff;
-            break;
-        case ADC_RES_12BIT:
-            adc_config[dev].max_value = 0xfff;
-            break;
-        case ADC_RES_14BIT:
-        case ADC_RES_16BIT:
-            adc_poweroff(dev);
-            return -1;
-            break;
-    }
-
-    /* set clock prescaler */
-    ADC->CCR = (3 << 16);       /* ADC clock = 10,5MHz */
-
-    /* enable the ADC module */
-    adc->CR2 |= ADC_CR2_ADON;
-
-    return 0;
+    return adc_config[line].dev;
 }
 
-int adc_sample(adc_t dev, int channel)
+static inline void poweron(adc_t line)
 {
-    ADC_TypeDef *adc = 0;
-
-    switch (dev) {
-#if ADC_0_EN
-        case ADC_0:
-            adc = ADC_0_DEV;
-            switch (channel) {
-                case 0:
-                    adc->SQR3 = ADC_0_CH0 & 0x1f;
-                    break;
-                case 1:
-                    adc->SQR3 = ADC_0_CH1 & 0x1f;
-                    break;
-                default:
-                    return -1;
-            }
-            break;
-#endif
-#if ADC_1_EN
-        case ADC_1:
-            adc = ADC_1_DEV;
-            switch (channel) {
-                case 0:
-                    adc->SQR3 = ADC_1_CH0 & 0x1f;
-                    break;
-                case 1:
-                    adc->SQR3 = ADC_1_CH1 & 0x1f;
-                    break;
-                default:
-                    return -1;
-            }
-            break;
-#endif
-    }
-
-    /* start single conversion */
-    adc->CR2 |= ADC_CR2_SWSTART;
-    /* wait until conversion is complete */
-    while (!(adc->SR & ADC_SR_EOC));
-    /* read and return result */
-    return (int)adc->DR;
+    RCC->APB2ENR |= (1 << adc_config[line].rcc);
```

Comment:
For the F(1|3)'s this is also the case, and probably also for the F0 and L series it's some sort of a business model :p